okpt/test/steps/opensearch.py

```diff
@@ -0,0 +1,208 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+
+#   http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Provides steps for OpenSearch tests.
+
+Some of the OpenSearch operations return a `took` field in the response body,
+so the profiling decorators aren't needed for some functions.
+"""
+from typing import Any, Dict, Iterable, List
+
+import elasticsearch
+import numpy as np
+
+from okpt.test import profile
+
+
+def bulk_partition_vectors(dataset: np.ndarray, split_size: int):
+    """Partitions an array of vectors into sections for bulk ingestion.
+
+    Partitions are formed by taking the first `n` vectors, then the next `n`,
+    and so on, until all the vectors have been assigned a partition group. The
+    last partition group may have less than `n` vectors.
+
+    Args:
+        dataset: Array of vectors to partition.
+        split_size: Size of each partition.
+
+    Returns:
+        An array of vector partitions.
+    """
+    splits = list(range(split_size, len(dataset), split_size))
+    return np.split(dataset[:], splits)
+
+
+def bulk_transform_vectors(dataset: np.ndarray, action: Dict[str, Any],
+                           bulk_size: int):
+    """Partitions and transforms a list of vectors into OpenSearch's bulk injection format.
+
+    Args:
+        dataset: An array of vectors to partition and transform.
+        action: The bulk action for each vector.
+        bulk_size: Number of vectors in a single bulk request.
+    Returns:
+        An array of partitioned and transformed vectors in bulk format.
+    """
+    def bulk_transform(section: np.ndarray):
+        """Helper function to transform vectors.
+
+        Args:
+            section: An array of vectors to transform.
+
+        Returns:
+            An array of transformed vectors in bulk format.
+        """
+        actions = [action, None] * len(section)
+        actions[1::2] = [{'test_vector': vec} for vec in section.tolist()]
+        return actions
+
+    partitions = bulk_partition_vectors(dataset, bulk_size)
+    partitions = map(bulk_transform, partitions)
+    return partitions
```

Comment:
So, if there are 1M vectors, all will be transformed and loaded into memory?

I think this wont scale.

Reply:
Good point, I've updated the transformation procedure to take place in chunks (each chunk is `bulk_size`).